doc/proposals/metering-operator-metrics.md

```diff
@@ -0,0 +1,99 @@
+## Auto register operator specific metrics as part of operator-metering
+
+### Motivation and goal
+
+We want to be able to generate the metering reports based on the operator specific Prometheus metrics. In order to be able to do that, operators must be instrumented to expose those metrics, and the operator-sdk should make this as easy as possible. The goal is to have the metering happen based on the usage of each individual operator. Metrics will be based on objects managed by the particular operator.
+
+### Overview of the metrics
+
+To follow both the Prometheus instrumentation [best practices](https://prometheus.io/docs/practices/naming/) as well as the official Kubernetes instrumentation [guide](https://github.com/kubernetes/community/blob/master/contributors/devel/instrumentation.md), the metrics will have the following format:
+
+```
+crd_kind_info{namespace="namespace",crdkind="instance-name"} 1
+```
+
+example metric for the memchached-operator would look like this:
+
+```
+memcached_info{namespace="default",memcached="example-memcached"} 1
+```
+
+### Existing solutions
+
+Either solutions makes use of Kubernetes list/watch to populate a Prometheus metrics registry, only that kube-state-metrics solution implements its own registry for performance reasons and the client_golang approach makes use of the canonical metrics registries.
+
+#### kube-state-metrics based
+
+[kube-state-metrics](https://github.com/kubernetes/kube-state-metrics#overview) can be used because it solves exactly the same problem but it does it for upstream known resources. The operator-sdk can re-use its functionality do perform the same thing with custom resources.
```

Comment:
The *kube-state-metrics library* can only be used for constant (/static) metrics, metrics that are immutable and thereby entirely regenerated on change. This is perfect for the above mentioned use-case.

It is not meant to do e.g. counting in performance critical code paths. Thereby a operator would need *kube-state-metrics library* for exposing the amount of custom resources that it manages and its details and Prometheus client_golang to expose metrics of its own internals e.g. count of reconciliation loops.

I think this is already clear to everyone, but could this be added to the proposal just to make sure?

Reply:
I didn't quite realize this and feel like what max wrote is probably worth being explicit about.